midp/fs.js

```diff
@@ -267,375 +263,463 @@ Native["com/sun/midp/rms/RecordStoreRegistry.stopAllRecordStoreListeners.(I)V"]
+    var pathname = util.fromJavaString(this.$nativePath);
+    DEBUG_FS && console.log("DefaultFileHandler.lastModified: " + pathname);
+    var stat = fs.stat(pathname);
+    return Long.fromNumber(stat != null ? stat.mtime : 0);
 };
 
-Native["com/ibm/oti/connection/file/FCInputStream.skipImpl.(JI)J"] = function(count, fd) {
-    var curpos = fs.getpos(fd);
-    var size = fs.getsize(fd);
-    if (curpos + count.toNumber() > size) {
-        fs.setpos(fd, size);
-        return Long.fromNumber(size - curpos);
+// A map from DefaultFileHandler instances with open input/output streams
+// to the file descriptors used by the fs API to identify open files.
+MIDP.openFileHandlers = new Map();
+
+Native["com/sun/cdc/io/j2me/file/DefaultFileHandler.openForRead.()V"] = function() {
+    var pathname = util.fromJavaString(this.$nativePath);
+    DEBUG_FS && console.log("DefaultFileHandler.openForRead: " + pathname);
+
+    if (MIDP.openFileHandlers.has(this)) {
+        // The file is already open, so we reset its position so the caller
+        // will start reading from the beginning of the file.
+        // XXX Should we track the position separately for the input and output
+        // streams?
+        var fd = MIDP.openFileHandlers.get(this);
+        fs.setpos(fd, 0);
+        return;
     }
 
-    fs.setpos(fd, curpos + count.toNumber());
-    return count;
-};
+    var stat = fs.stat(pathname);
 
-Native["com/ibm/oti/connection/file/FCInputStream.readImpl.([BIII)I"] = function(buffer, offset, count, fd) {
-    if (offset < 0 || count < 0 || offset > buffer.byteLength || (buffer.byteLength - offset) < count) {
-        throw $.newIndexOutOfBoundsException();
+    if (!stat) {
+        throw $.newIOException("file doesn't exist");
     }
 
-    if (buffer.byteLength == 0 || count == 0) {
-        return 0;
+    if (stat.isDir) {
+        throw $.newIOException("file is a directory");
     }
 
-    var curpos = fs.getpos(fd);
-    var data = fs.read(fd, curpos, curpos + count);
-    buffer.set(data, offset);
+    var fileHandler = this;
 
-    return (data.byteLength > 0) ? data.byteLength : -1;
+    asyncImpl("I", new Promise(function(resolve, reject) {
+        fs.open(pathname, function(fd) {
+            MIDP.openFileHandlers.set(fileHandler, fd);
+            resolve();
+        });
+    }));
 };
 
-Native["com/ibm/oti/connection/file/FCInputStream.readByteImpl.(I)I"] = function(fd) {
-    var curpos = fs.getpos(fd);
+Native["com/sun/cdc/io/j2me/file/DefaultFileHandler.closeForRead.()V"] = function() {
+    var pathname = util.fromJavaString(this.$nativePath);
+    DEBUG_FS && console.log("DefaultFileHandler.closeForRead: " + pathname);
+    var fd = MIDP.openFileHandlers.get(this);
+    fs.close(fd);
+    MIDP.openFileHandlers.delete(this);
+};
 
-    var data = fs.read(fd, curpos, curpos+1);
+Native["com/sun/cdc/io/j2me/file/DefaultFileHandler.openForWrite.()V"] = function() {
```

Comment:
This is similar (or equal?) to openForRead, could you refactor the code in a single function?